pkg/cmd/provisioning/azure/create_managed_identities.go

```diff
@@ -0,0 +1,467 @@
+			err := assignRoleToManagedIdentity(client, *userAssignedManagedIdentity.Properties.PrincipalID, roleBinding.Role, scope, subscriptionID)
+			if err != nil {
+				return errors.Wrapf(err, "failed to assign role %s to user-assigned managed identity", roleBinding.Role)
+			}
+		}
+	}
+
+	// Create a federated identity credential for every service account enumerated in the CredentialsRequest
+	for _, serviceAccountName := range credentialsRequest.Spec.ServiceAccountNames {
+		err := createFederatedIdentityCredential(client, shortenedManagedIdentityName, issuerURL, credentialsRequest.Spec.SecretRef.Namespace, serviceAccountName, resourceGroupName)
+		if err != nil {
+			return err
+		}
+	}
+
+	writeCredReqSecret(credentialsRequest, outputDir, *userAssignedManagedIdentity.Properties.ClientID, *userAssignedManagedIdentity.Properties.TenantID, subscriptionID, region)
+	return nil
+}
+
+// getRoleDefinition lists role definitions within the scope of the Azure subscription identified by subscriptionID
+// and returns the armauthorization.RoleDefinition with a name matching the provided roleName.
+//
+// If multiple roles are found matching the roleName this will result in an error.
+func getRoleDefinition(client *azureclients.AzureClientWrapper, roleName, subscriptionID string) (*armauthorization.RoleDefinition, error) {
+	listRoles := client.RoleDefinitionsClient.NewListPager(
+		"/subscriptions/"+subscriptionID,
+		&armauthorization.RoleDefinitionsClientListOptions{
+			Filter: to.Ptr(fmt.Sprintf("roleName eq '%v'", roleName)),
+		},
+	)
+	roleDefinitions := make([]*armauthorization.RoleDefinition, 0)
+	for listRoles.More() {
+		pageResponse, err := listRoles.NextPage(context.TODO())
+		if err != nil {
+			return nil, err
+		}
+		roleDefinitions = append(roleDefinitions, pageResponse.RoleDefinitionListResult.Value...)
+	}
+	switch len(roleDefinitions) {
+	case 0:
+		return nil, fmt.Errorf("no role found for name %q", roleName)
+	case 1:
+		return roleDefinitions[0], nil
+	default:
+		return nil, fmt.Errorf("found %q role definitions for %q, expected one", len(roleDefinitions), roleName)
+	}
+}
+
+// assignRoleToManagedIdentity assigns the Azure role with roleName within the provided subscriptionID to the
+// managed identity identified by managedIdentityPrincipalID within the provided scope.
+//
+// Scope is a string such as /subscriptions/<subscriptionID> which represents anything within the subscription.
+// This scope can be restricted within a resourceGroup such as /subscriptions/<subscriptionID>/resourceGroups/<resourceGroupName>.
+func assignRoleToManagedIdentity(client *azureclients.AzureClientWrapper, managedIdentityPrincipalID, roleName, scope, subscriptionID string) error {
+	targetRole, err := getRoleDefinition(client, roleName, subscriptionID)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to get role definition for role %s", roleName))
+	}
+
+	uuid, err := uuid.NewV4()
```

Comment:
Code review: is `NewV4()` the right uuid generation to use here and what are the chances for collisions which would result in a bounce from Azure?

Reply:
ref: [RFC-4122](https://www.rfc-editor.org/rfc/rfc4122.txt). I'm going to swap the library we use to generate the UUID to  https://github.com/google/uuid which we're using elsewhere in CCO. I'm not sure what our chances of collision are or if this is insufficient. Worst case, I think it would be rare to encounter a dupe and we'd have to accept that this would mean the user has to re-run `create-managed-identities` which should be idempotent.

Reply:
[Done](https://github.com/openshift/cloud-credential-operator/pull/523/commits/62028d2f74e1b5c97de89c9486595b4ab3e6f9be)